ompi/mca/coll/base/coll_base_allreduce.c

```diff
@@ -1321,17 +1326,25 @@ int ompi_coll_base_allreduce_intra_allgather_reduce(const void *sbuf, void *rbuf
 
     err = ompi_request_wait_all(req_index, reqs, MPI_STATUSES_IGNORE);
 
-    /* Prepare for local reduction */
-    peer_rank = 0;
-    if (!commutative) {
-        /* For non-commutative operations, ensure the reduction always starts from Rank 0's data */
-        memcpy(rbuf, 0 == rank ? send_buf : tmp_buf, incr);
-        peer_rank = 1;
+    /**
+     * Prepare for local reduction:
+     * 1. In the previous exchange we did not send data to self - do that right now
+     * 2. Ensure the reduction always starts from Rank 0's data
+     */
+    if (0 != rank && rbuf == send_buf) {
```

Comment:
I had to read the if-block lines several times, I think it was this preceding comment that tripped me up.  In particular, the comment claims we will send to self right now, but actually we only do so in a very specific case (in-place), when not rank 0.  Can we change the comment to something like:

> When performing in-place reduction, we must not overwrite our input buffer while we reduce in-order.  Rank 0 is an exception, since the reduction starts there.

Reply:
After eating food, I realize that I missed a case for `0 == rank && rbuf != send_buf` -  we still need to copy the data in that case.

Please see update.